### ISSUE TITLE
Multithreading Issue

### ISSUE THREAD

**Issue by Pebaz** [Pebaz/nimporter#19]

**Comment by Benjamin-Lee** [comment]
For reference, I have this solved using 

```nim
# testing.nim
import nimpy

proc multiplier*(a: int, b: int): int {.exportpy.} =
    var x = 0
    echo a, " ",  b

    # noop to check CPU usage
    for i in 0..100000000000:
        x = x + a + b
        
    return a * b
```

Python code:

```py
import multiprocessing
import testing


# square function
def square(x):
    return testing.multiplier(
        x,
        x,
    )


if __name__ == "__main__":

    # multiprocessing pool object
    pool = multiprocessing.Pool()

    # pool object with number of element
    pool = multiprocessing.Pool(processes=6)

    # input list
    inputs = [0, 1, 2, 3, 4]

    # map the function to the list and pass
    # function and input list as arguments
    outputs = pool.map(square, inputs)

    # Print input list
    print("Input: {}".format(inputs))

    # Print output list
    print("Output: {}".format(outputs))
```

Output:

```
$ python test.py
0 0
1 1
2 2
3 3
4 4
Input: [0, 1, 2, 3, 4]
Output: [0, 1, 4, 9, 16]
```

Proof it works:

<img width="852" alt="image" src="https://user-images.githubusercontent.com/13618205/157554351-7d94aa5f-dd66-4aac-9b75-8efd4ed71820.png">

**Comment by Iapetus-11** [comment]
Any update on this?

**Comment by Pebaz** [comment]
Hi @Iapetus-11, from what I understand, this doesn't affect Linux users and there's a workaround on MacOS listed above, but Windows still suffers from this I believe. Since most people will run Flask from Linux VMs or containers, hopefully this is not too big of an issue.

**Comment by Iapetus-11** [comment]
- The workaround is not a workaround, multiprocessing is very different than threading in Python. Multiprocessing spins up new Python interpreters whereas threading does not and code execution in threads is limited by the GIL.
   - Also, using multiprocessing with nimporter does actually appear to work on windows, probably because it's seperate Py interpreters. Test code and output here: https://gist.github.com/Iapetus-11/2ae39ef8d490546771d3550b0b79cbdc
- Also, if I change the usage of `multiprocessing.Pool` to something like `ThreadPoolExecutor` or if I even just use `threading.Thread`, when calling nim code, my program crashes with no output or any indication of error. I did notice that the errorlevel variable in command prompt is set to `-1073741819` after the crash, if that helps at all.
- I have mostly moved to docker and mac but it's nice to write code which works everywhere.

Anyways, thank you for writing such a great library. If you need any help testing or need any additional information from my tests please let me know.

**Comment by Pebaz** [comment]
True that, yeah `multiprocessing` works since it's presumably using IPC with different processes so they each load their own version of the Nim extension dynamic library. I'm sure this has to do with executable memory pages on Windows but I'm not sure if it's something that needs to be set when Python is compiled or a Python CLI flag or even a Nim compile flag. I need to get with the Nimpy guy again and see if he has any ideas on how to proceed. 😁

**Comment by Pebaz** [comment]
@Iapetus-11, if you're interested, I've got an in-progress complete rewrite of Nimporter that I've needed someone to help test: https://github.com/Pebaz/nimporter/releases/tag/v2.0.0rc

The reason I haven't merged it so far is because it diverges from the current implementation of Nimporter v1 significantly but in return, it solidifies on the original idea of Nimporter which was to make Python + Nim projects dead-simple.

Basically, I'm looking for reasons why it *wouldn't work*. One example is that pure-Nim libraries are the easiest to implement but what about wrapping something like Raylib with Nim and exposing it to Python? Would that work given the new v2 semantics on all platforms? Also note that the platforms listed in v2 need to be added onto, I just wanted to cover Win32, MacOS, and Linux initially.

**Comment by Iapetus-11** [comment]
> I need to get with the Nimpy guy again and see if he has any ideas on how to proceed. 😁

That'd be awesome

> Basically, I'm looking for reasons why it wouldn't work. One example is that pure-Nim libraries are the easiest to implement but what about wrapping something like Raylib with Nim and exposing it to Python? Would that work given the new v2 semantics on all platforms? Also note that the platforms listed in v2 need to be added onto, I just wanted to cover Win32, MacOS, and Linux initially.

I'm not familiar with RayLib but I do have a project which uses nimpy's raw py buffers api and several other projects that use pure nim (and nimporter) I could test with later today.
